### PR TITLE
feat(testing): mempool tests and helper functions

### PR DESCRIPTION
**Pull request by shampoobera**
todo

- [ ] possibly separate concurrency tests into `mempool_race_test.go`
- [ ] fix slow code

### CODE REVIEW

File: cosmos/x/evm/plugins/txpool/mempool/mempool_test.go [hunk 1]
```diff
@@ -155,6 +180,32 @@ func (mplf *mockPLF) Build(event *sdk.Event) (*coretypes.Log, error) {
...
+		for _, ethTx := range list {
+			if tx.Hash() == ethTx.Hash() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isPendingTx(mempool *EthTxPool, tx *coretypes.Transaction) bool {
```
Comment (itsdevbear) [inline cosmos/x/evm/plugins/txpool/mempool/mempool_test.go:196]: make the mempool the receiver imo

Comment (itsdevbear) [reply cosmos/x/evm/plugins/txpool/mempool/mempool_test.go:196]: these are lowkey pretty inefficient but thats okay for now

Comment (shampoobera) [reply cosmos/x/evm/plugins/txpool/mempool/mempool_test.go]: ya i'll optimize just wanted to get it working

**Comment by itsdevbear** [comment]
@shampoobera call it "race" 

Go terminology

File: cosmos/x/evm/plugins/txpool/mempool/mempool_test.go [hunk 2]
```diff
@@ -136,6 +138,99 @@ var _ = Describe("EthTxPool", func() {
...
+				Expect(etp.isQueuedTx(etp, ethtx3)).To(BeTrue())
+
+				_, tx2 := buildTx(key1, &coretypes.LegacyTx{Nonce: 2})
+				Expect(etp.Insert(ctx, tx2)).ToNot(HaveOccurred())
+
+				_, queuedTransactions := etp.ContentFrom(addr1)
+				Expect(queuedTransactions).To(HaveLen(0))
+				Expect(etp.Nonce(addr1)).To(Equal(uint64(4)))
+			})
+		It("should not allow duplicate nonces (replay attack)", func() {
```
Comment (itsdevbear) [inline cosmos/x/evm/plugins/txpool/mempool/mempool_test.go:158]: Should allow if the gas is increased by 10% tho!

File: cosmos/x/evm/plugins/txpool/mempool/mempool_test.go [hunk 3]
```diff
@@ -136,6 +140,147 @@ var _ = Describe("EthTxPool", func() {
...
+			// ... and peer B
+			go func(etp *EthTxPool) {
+				for _, list := range etp.senderIndices {
+					for elem := list.Front(); elem != nil; elem = elem.Next() {
+						readsFromB++
+					}
+				}
+			}(etp)
+
+			Expect(readsFromA).To(BeEquivalentTo(readsFromB))
```
Comment (itsdevbear) [inline cosmos/x/evm/plugins/txpool/mempool/mempool_test.go:282]: Probably want to use a wait group here https://gobyexample.com/waitgroups

**Comment by itsdevbear** [comment]
looking fire chief

**Comment by itsdevbear** [comment]
📸🐻

**Review by calbera — APPROVED** [review]
inshallah we have a mempool

**Comment by calbera** [comment]
CodeCov
image

**Comment by itsdevbear** [comment]
this definitely still has the ddos issue, but merging into the mega branch for now.

**Comment by itsdevbear** [comment]
good stuff @shampoobera :)))

**Pull request merged**